// @cred/neopop-web@1.0.0/lib/components/Header/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../_virtual/_rollupPluginBabelHelpers.js');
var React = require('react');
require('../../primitives/index.js');
var index$1 = require('../Back/index.js');
var index = require('../Helpers/index.js');
var index$2 = require('../Typography/index.js');
var typography = require('../../primitives/typography.js');
var colors = require('../../primitives/colors.js');
var opacity = require('../../primitives/opacity.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var Header = function Header(props) {
  var header = typography.typographyGuide.header;
  var heading = props.heading,
      _props$textStyle = props.textStyle,
      textStyle = _props$textStyle === void 0 ? {
    heading: header.heading,
    description: header.description
  } : _props$textStyle,
      color = props.color,
      description = props.description,
      onBackClick = props.onBackClick;
  var fallbackColor = color !== null && color !== void 0 ? color : colors.mainColors.white;
  return /*#__PURE__*/React__default["default"].createElement(index.Column, null, /*#__PURE__*/React__default["default"].createElement(index$1["default"], {
    onClick: onBackClick,
    color: fallbackColor
  }), heading ? /*#__PURE__*/React__default["default"].createElement("div", {
    style: {
      paddingTop: '10px'
    }
  }, /*#__PURE__*/React__default["default"].createElement(index$2["default"], _rollupPluginBabelHelpers["extends"]({}, textStyle.heading, {
    style: {
      maxWidth: '75%',
      opacity: opacity.FontOpacity.HEADING
    },
    color: fallbackColor
  }), heading)) : '', description ? /*#__PURE__*/React__default["default"].createElement("div", {
    style: {
      paddingTop: '5px'
    }
  }, /*#__PURE__*/React__default["default"].createElement(index$2["default"], _rollupPluginBabelHelpers["extends"]({}, textStyle.description, {
    style: {
      maxWidth: '75%',
      opacity: opacity.FontOpacity.BODY_TEXT
    },
    color: fallbackColor
  }), description)) : '');
};

exports["default"] = Header;
