// @cred/neopop-web@1.0.0/lib/components/Slider/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var styles = require('./styles.js');
require('../../primitives/index.js');
var index = require('../../utils/index.js');
var slider = require('../../primitives/slider.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var Slider = function Slider(props) {
  var min = props.min,
      max = props.max,
      step = props.step,
      defaultValue = props.defaultValue,
      value = props.value,
      onInput = props.onInput,
      onChange = props.onChange,
      style = props.style,
      sliderConfig = props.sliderConfig,
      thumbConfig = props.thumbConfig;
  var sliderUiConfig = index.isEmpty(sliderConfig) ? slider.SliderConfig : sliderConfig;
  var thumbUiConfig = index.isEmpty(thumbConfig) ? slider.ThumbConfig : thumbConfig;
  return /*#__PURE__*/React__default["default"].createElement(styles.SliderWrapper, {
    sliderConfig: sliderUiConfig,
    thumbConfig: thumbUiConfig
  }, /*#__PURE__*/React__default["default"].createElement("input", {
    className: "slider",
    type: "range",
    min: min,
    max: max,
    step: step,
    defaultValue: defaultValue,
    value: value,
    onInput: onInput,
    onChange: onChange,
    style: style
  }));
};

exports["default"] = Slider;
