// @cred/neopop-web@1.0.0/lib-esm/components/BottomSheet/index.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends, slicedToArray as _slicedToArray } from '../../_virtual/_rollupPluginBabelHelpers.js';
import React__default, { useState, useLayoutEffect, useEffect } from 'react';
import BottomSheetControlled, { BottomSheetControlledCustom } from './components/BottomSheetControlled.js';

var _excluded = ["children", "onAfterClose"],
    _excluded2 = ["children", "onAfterClose"];
var useBottomSheet = function useBottomSheet(_ref) {
  var onAfterClose = _ref.onAfterClose,
      open = _ref.open;

  var _useState = useState(false),
      _useState2 = _slicedToArray(_useState, 2),
      mounted = _useState2[0],
      setMounted = _useState2[1];

  var handleCloseEnd = function handleCloseEnd() {
    if (onAfterClose instanceof Function) {
      onAfterClose();
    }

    setMounted(false);
  };

  useLayoutEffect(function () {
    if (open) {
      setMounted(true);
    }
  }, [open]);
  useEffect(function () {
    if (open) {
      document.body.style.overflow = 'hidden';
    }

    return function () {
      document.body.style.overflow = 'visible';
    };
  }, [open]);
  return {
    mounted: mounted,
    handleCloseEnd: handleCloseEnd
  };
};

var BottomSheet = function BottomSheet(_ref2) {
  var children = _ref2.children,
      onAfterClose = _ref2.onAfterClose,
      props = _objectWithoutProperties(_ref2, _excluded);

  var _useBottomSheet = useBottomSheet({
    onAfterClose: onAfterClose,
    open: props.open
  }),
      handleCloseEnd = _useBottomSheet.handleCloseEnd,
      mounted = _useBottomSheet.mounted;

  if (!mounted) {
    return null;
  }

  return /*#__PURE__*/React__default.createElement(BottomSheetControlled, _extends({
    onCloseEnd: handleCloseEnd
  }, props), children);
};

var BottomSheetCustom = function BottomSheetCustom(_ref3) {
  var children = _ref3.children,
      onAfterClose = _ref3.onAfterClose,
      props = _objectWithoutProperties(_ref3, _excluded2);

  var _useBottomSheet2 = useBottomSheet({
    onAfterClose: onAfterClose,
    open: props.open
  }),
      handleCloseEnd = _useBottomSheet2.handleCloseEnd,
      mounted = _useBottomSheet2.mounted;

  if (!mounted) {
    return null;
  }

  return /*#__PURE__*/React__default.createElement(BottomSheetControlledCustom, _extends({
    onCloseEnd: handleCloseEnd
  }, props), children);
};

export { BottomSheetCustom, BottomSheet as default, useBottomSheet };
