// @cred/neopop-web@1.0.0/lib-esm/components/BottomSheet/styles.js
import styled from 'styled-components';
import { a as animated } from '../../node_modules/@react-spring/web/dist/react-spring-web.esm.js';
import { hexToRGBA } from '../../utils/index.js';
import '../../primitives/index.js';
import { mainColors, colorPalette } from '../../primitives/colors.js';

var Wraper = styled.div.withConfig({
  displayName: "styles__Wraper",
  componentId: "sc-b2exit-0"
})(["position:relative;z-index:101;"]);
var Overlay = styled(animated.div).withConfig({
  displayName: "styles__Overlay",
  componentId: "sc-b2exit-1"
})(["position:fixed;top:-60px;right:0px;left:0px;bottom:-60px;background-color:", ";touch-action:none;z-index:5;"], hexToRGBA(mainColors.black, 0.9));
var Sheet = styled(animated.div).withConfig({
  displayName: "styles__Sheet",
  componentId: "sc-b2exit-2"
})(["width:100%;overflow:auto;overscroll-behavior:auto;scrollbar-width:none;width:100%;&::-webkit-scrollbar{width:0px;background:transparent;}"]);
var CardCustom = styled(animated.div).withConfig({
  displayName: "styles__CardCustom",
  componentId: "sc-b2exit-3"
})(["display:flex;position:fixed;left:0;bottom:0;right:0;width:100%;max-height:", ";z-index:6;touch-action:none;"], function (p) {
  return p.$maxHeight;
});
var Card = styled(CardCustom).withConfig({
  displayName: "styles__Card",
  componentId: "sc-b2exit-4"
})(["max-height:", ";background-color:", ";&::after,&::before{position:absolute;content:' ';top:-3.5px;height:4px;width:55%;background-color:", ";z-index:1;}&::after{left:0;margin-left:1.5px;transform:skewX(-45deg);}&::before{right:0;margin-right:1.5px;transform:skewX(45deg);}"], function (p) {
  return p.$maxHeight;
}, colorPalette.popWhite[400], function (p) {
  return p.$sheetShadowColor ? p.$sheetShadowColor : '#c4c4c4';
});
var Notch = styled.div.withConfig({
  displayName: "styles__Notch",
  componentId: "sc-b2exit-5"
})(["position:absolute;top:-18px;width:60px;height:2px;background-color:", ";left:50%;transform:translate(-50%,0);"], colorPalette.popBlack[100]);
var Content = styled.div.withConfig({
  displayName: "styles__Content",
  componentId: "sc-b2exit-6"
})(["&::after,&::before{content:' ';clear:both;display:table;}"]);
var GapFiller = styled(animated.div).withConfig({
  displayName: "styles__GapFiller",
  componentId: "sc-b2exit-7"
})(["background-color:", ";height:1px;display:block;transform-origin:bottom;z-index:10;overscroll-behavior:none;touch-action:none;position:fixed;right:0;bottom:0;left:0;-webkit-tap-highlight-color:transparent;-webkit-touch-callout:none;"], colorPalette.popWhite[400]);

export { Card, CardCustom, Content, GapFiller, Notch, Overlay, Sheet, Wraper };
