// @cred/neopop-web@1.0.0/lib-esm/components/Back/index.js
import { objectSpread2 as _objectSpread2, extends as _extends } from '../../_virtual/_rollupPluginBabelHelpers.js';
import React__default from 'react';
import '../../primitives/index.js';
import { Row, VerticalSpacer } from '../Helpers/index.js';
import Typography from '../Typography/index.js';
import { typographyGuide } from '../../primitives/typography.js';
import { mainColors } from '../../primitives/colors.js';

var Back = function Back(props) {
  var heading = props.heading,
      color = props.color,
      _props$textStyle = props.textStyle,
      textStyle = _props$textStyle === void 0 ? _objectSpread2({}, typographyGuide.back.heading) : _props$textStyle,
      rightElement = props.rightElement,
      onClick = props.onClick;
  return /*#__PURE__*/React__default.createElement(Row, {
    className: "h-center",
    style: {
      padding: '12px 0px'
    }
  }, /*#__PURE__*/React__default.createElement(Row, {
    className: "h-center"
  }, /*#__PURE__*/React__default.createElement(Row, {
    className: "h-center",
    style: {
      height: '36px'
    },
    onClick: onClick
  }, /*#__PURE__*/React__default.createElement("svg", {
    width: "32",
    height: "12",
    viewBox: "0 0 32 12",
    fill: "none"
  }, /*#__PURE__*/React__default.createElement("path", {
    d: "M31 6L0.999999 6",
    stroke: color !== null && color !== void 0 ? color : mainColors.white,
    strokeWidth: "1.5",
    strokeLinecap: "square",
    strokeLinejoin: "round"
  }), /*#__PURE__*/React__default.createElement("path", {
    d: "M7 12C7 8.68629 4.31371 6 1 6",
    stroke: color !== null && color !== void 0 ? color : mainColors.white,
    strokeWidth: "1.5"
  }), /*#__PURE__*/React__default.createElement("path", {
    d: "M1 6C4.31371 6 7 3.31371 7 0",
    stroke: color !== null && color !== void 0 ? color : mainColors.white,
    strokeWidth: "1.5"
  }))), /*#__PURE__*/React__default.createElement(VerticalSpacer, {
    n: 4
  }), heading ? /*#__PURE__*/React__default.createElement(Typography, _extends({}, textStyle, {
    color: color !== null && color !== void 0 ? color : mainColors.white
  }), heading) : null), /*#__PURE__*/React__default.createElement("div", {
    style: {
      marginLeft: 'auto'
    }
  }, rightElement !== null && rightElement !== void 0 ? rightElement : null));
};

export { Back as default };
