// @cred/neopop-web@1.0.0/lib/components/BottomSheet/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../_virtual/_rollupPluginBabelHelpers.js');
var React = require('react');
var BottomSheetControlled = require('./components/BottomSheetControlled.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var _excluded = ["children", "onAfterClose"],
    _excluded2 = ["children", "onAfterClose"];
var useBottomSheet = function useBottomSheet(_ref) {
  var onAfterClose = _ref.onAfterClose,
      open = _ref.open;

  var _useState = React.useState(false),
      _useState2 = _rollupPluginBabelHelpers.slicedToArray(_useState, 2),
      mounted = _useState2[0],
      setMounted = _useState2[1];

  var handleCloseEnd = function handleCloseEnd() {
    if (onAfterClose instanceof Function) {
      onAfterClose();
    }

    setMounted(false);
  };

  React.useLayoutEffect(function () {
    if (open) {
      setMounted(true);
    }
  }, [open]);
  React.useEffect(function () {
    if (open) {
      document.body.style.overflow = 'hidden';
    }

    return function () {
      document.body.style.overflow = 'visible';
    };
  }, [open]);
  return {
    mounted: mounted,
    handleCloseEnd: handleCloseEnd
  };
};

var BottomSheet = function BottomSheet(_ref2) {
  var children = _ref2.children,
      onAfterClose = _ref2.onAfterClose,
      props = _rollupPluginBabelHelpers.objectWithoutProperties(_ref2, _excluded);

  var _useBottomSheet = useBottomSheet({
    onAfterClose: onAfterClose,
    open: props.open
  }),
      handleCloseEnd = _useBottomSheet.handleCloseEnd,
      mounted = _useBottomSheet.mounted;

  if (!mounted) {
    return null;
  }

  return /*#__PURE__*/React__default["default"].createElement(BottomSheetControlled["default"], _rollupPluginBabelHelpers["extends"]({
    onCloseEnd: handleCloseEnd
  }, props), children);
};

var BottomSheetCustom = function BottomSheetCustom(_ref3) {
  var children = _ref3.children,
      onAfterClose = _ref3.onAfterClose,
      props = _rollupPluginBabelHelpers.objectWithoutProperties(_ref3, _excluded2);

  var _useBottomSheet2 = useBottomSheet({
    onAfterClose: onAfterClose,
    open: props.open
  }),
      handleCloseEnd = _useBottomSheet2.handleCloseEnd,
      mounted = _useBottomSheet2.mounted;

  if (!mounted) {
    return null;
  }

  return /*#__PURE__*/React__default["default"].createElement(BottomSheetControlled.BottomSheetControlledCustom, _rollupPluginBabelHelpers["extends"]({
    onCloseEnd: handleCloseEnd
  }, props), children);
};

exports.BottomSheetCustom = BottomSheetCustom;
exports["default"] = BottomSheet;
exports.useBottomSheet = useBottomSheet;
