// @cred/neopop-web@1.0.0/lib/primitives/scoremeter.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var colors = require('./colors.js');

exports.Bureaus = void 0;

(function (Bureaus) {
  Bureaus["crif"] = "crif";
  Bureaus["experian"] = "experian";
  Bureaus["equifax"] = "equifax";
})(exports.Bureaus || (exports.Bureaus = {}));

var scoreConfig = {
  excellent: {
    type: 'excellent',
    text: 'Excellent'
  },
  average: {
    type: 'average',
    text: 'Good'
  },
  poor: {
    type: 'poor',
    text: 'Poor'
  }
};
var getTypeAndText = function getTypeAndText(score, bureau) {
  if (bureau === exports.Bureaus.experian && score < 850) {
    return score >= 750 ? scoreConfig.average : scoreConfig.poor;
  } else if (bureau === exports.Bureaus.crif && score < 770) {
    return score >= 600 ? scoreConfig.average : scoreConfig.poor;
  } else if (score < 780) {
    return score >= 690 ? scoreConfig.average : scoreConfig.poor;
  }

  return scoreConfig.excellent;
};

var darkScoreMeter = function darkScoreMeter(score, bureau) {
  return {
    colorMode: 'dark',
    reading: score,
    colorConfig: colors.colorGuide.darkComponents.scoremeter,
    type: getTypeAndText(score, bureau).type,
    scoreDesc: getTypeAndText(score, bureau).text
  };
};

var lightScoreMeter = function lightScoreMeter(score, bureau, oldScore) {
  return {
    colorMode: 'light',
    reading: score,
    oldReading: oldScore,
    colorConfig: colors.colorGuide.lightComponents.scoremeter,
    type: getTypeAndText(score, bureau).type,
    showIndicators: true,
    scoreDesc: getTypeAndText(score, bureau).text
  };
};

var getScoreMeterConfig = {
  lightScoreMeter: lightScoreMeter,
  darkScoreMeter: darkScoreMeter
};

exports.getScoreMeterConfig = getScoreMeterConfig;
exports.getTypeAndText = getTypeAndText;
