// @cred/neopop-web@1.0.0/lib/primitives/typography.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../_virtual/_rollupPluginBabelHelpers.js');
require('../components/Button/types.js');
var types = require('../components/Typography/types.js');
require('../components/Helpers/types.js');

var fontNameSpaces = _rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2({}, [44, 34, 28, 22, 20, 18, 16, 15, 14, 13].reduce(function (accum, fontSize) {
  return _rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2({}, accum), {}, _rollupPluginBabelHelpers.defineProperty({}, "th".concat(fontSize, "eb"), {
    fontType: types.FontType.HEADING,
    fontSize: fontSize,
    fontWeight: types.FontWeights.EXTRA_BOLD
  }));
}, {})), [34, 28, 22, 20, 18, 16, 14, 13, 11].reduce(function (accum, fontSize) {
  return _rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2({}, accum), {}, _rollupPluginBabelHelpers.defineProperty({}, "th".concat(fontSize, "b"), {
    fontType: types.FontType.HEADING,
    fontSize: fontSize,
    fontWeight: types.FontWeights.BOLD
  }));
}, {})), [22, 20, 18, 16, 14, 13, 12, 10].reduce(function (accum, fontSize) {
  return _rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2({}, accum), {}, _rollupPluginBabelHelpers.defineProperty({}, "th".concat(fontSize, "sb"), {
    fontType: types.FontType.HEADING,
    fontSize: fontSize,
    fontWeight: types.FontWeights.SEMI_BOLD
  }));
}, {})), [16, 15, 14, 13, 12, 11].reduce(function (accum, fontSize) {
  return _rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2({}, accum), {}, _rollupPluginBabelHelpers.defineProperty({}, "tb".concat(fontSize, "m"), {
    fontType: types.FontType.BODY,
    fontSize: fontSize,
    fontWeight: types.FontWeights.MEDIUM
  }));
}, {})), [16, 15, 14, 13, 12, 11].reduce(function (accum, fontSize) {
  return _rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2({}, accum), {}, _rollupPluginBabelHelpers.defineProperty({}, "tb".concat(fontSize, "r"), {
    fontType: types.FontType.BODY,
    fontSize: fontSize,
    fontWeight: types.FontWeights.REGULAR
  }));
}, {})), {}, {
  tc12eb: {
    fontType: types.FontType.CAPS,
    fontSize: 12,
    fontWeight: types.FontWeights.EXTRA_BOLD
  },
  tc10eb: {
    fontType: types.FontType.CAPS,
    fontSize: 10,
    fontWeight: types.FontWeights.EXTRA_BOLD
  },
  tc12b: {
    fontType: types.FontType.CAPS,
    fontSize: 12,
    fontWeight: types.FontWeights.BOLD
  },
  tc10b: {
    fontType: types.FontType.CAPS,
    fontSize: 10,
    fontWeight: types.FontWeights.BOLD
  },
  tc8b: {
    fontType: types.FontType.CAPS,
    fontSize: 8,
    fontWeight: types.FontWeights.BOLD
  },
  tc12sb: {
    fontType: types.FontType.CAPS,
    fontSize: 12,
    fontWeight: types.FontWeights.SEMI_BOLD
  },
  tc10sb: {
    fontType: types.FontType.CAPS,
    fontSize: 10,
    fontWeight: types.FontWeights.SEMI_BOLD
  }
}, [36, 34, 32, 24, 22, 20, 18].reduce(function (accum, fontSize) {
  return _rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2({}, accum), {}, _rollupPluginBabelHelpers.defineProperty({}, "tsh".concat(fontSize, "b"), {
    fontType: types.FontType.SERIF_HEADING,
    fontSize: fontSize,
    fontWeight: types.FontWeights.BOLD
  }));
}, {}));
var FontVariant = {
  HeadingExtraBold44: {
    fontType: types.FontType.HEADING,
    fontSize: 44,
    fontWeight: types.FontWeights.EXTRA_BOLD
  },
  HeadingExtraBold34: {
    fontType: types.FontType.HEADING,
    fontSize: 34,
    fontWeight: types.FontWeights.EXTRA_BOLD
  },
  HeadingExtraBold28: {
    fontType: types.FontType.HEADING,
    fontSize: 28,
    fontWeight: types.FontWeights.EXTRA_BOLD
  },
  HeadingExtraBold22: {
    fontType: types.FontType.HEADING,
    fontSize: 22,
    fontWeight: types.FontWeights.EXTRA_BOLD
  },
  HeadingExtraBold20: {
    fontType: types.FontType.HEADING,
    fontSize: 20,
    fontWeight: types.FontWeights.EXTRA_BOLD
  },
  HeadingExtraBold18: {
    fontType: types.FontType.HEADING,
    fontSize: 18,
    fontWeight: types.FontWeights.EXTRA_BOLD
  },
  HeadingExtraBold16: {
    fontType: types.FontType.HEADING,
    fontSize: 16,
    fontWeight: types.FontWeights.EXTRA_BOLD
  },
  HeadingExtraBold15: {
    fontType: types.FontType.HEADING,
    fontSize: 15,
    fontWeight: types.FontWeights.EXTRA_BOLD
  },
  HeadingExtraBold14: {
    fontType: types.FontType.HEADING,
    fontSize: 14,
    fontWeight: types.FontWeights.EXTRA_BOLD
  },
  HeadingExtraBold13: {
    fontType: types.FontType.HEADING,
    fontSize: 13,
    fontWeight: types.FontWeights.EXTRA_BOLD
  },
  HeadingBold34: {
    fontType: types.FontType.HEADING,
    fontSize: 34,
    fontWeight: types.FontWeights.BOLD
  },
  HeadingBold28: {
    fontType: types.FontType.HEADING,
    fontSize: 28,
    fontWeight: types.FontWeights.BOLD
  },
  HeadingBold22: {
    fontType: types.FontType.HEADING,
    fontSize: 22,
    fontWeight: types.FontWeights.BOLD
  },
  HeadingBold20: {
    fontType: types.FontType.HEADING,
    fontSize: 20,
    fontWeight: types.FontWeights.BOLD
  },
  HeadingBold18: {
    fontType: types.FontType.HEADING,
    fontSize: 18,
    fontWeight: types.FontWeights.BOLD
  },
  HeadingBold16: {
    fontType: types.FontType.HEADING,
    fontSize: 16,
    fontWeight: types.FontWeights.BOLD
  },
  HeadingBold14: {
    fontType: types.FontType.HEADING,
    fontSize: 14,
    fontWeight: types.FontWeights.BOLD
  },
  HeadingBold13: {
    fontType: types.FontType.HEADING,
    fontSize: 13,
    fontWeight: types.FontWeights.BOLD
  },
  HeadingBold11: {
    fontType: types.FontType.HEADING,
    fontSize: 11,
    fontWeight: types.FontWeights.BOLD
  },
  HeadingSemiBold22: {
    fontType: types.FontType.HEADING,
    fontSize: 22,
    fontWeight: types.FontWeights.SEMI_BOLD
  },
  HeadingSemiBold20: {
    fontType: types.FontType.HEADING,
    fontSize: 20,
    fontWeight: types.FontWeights.SEMI_BOLD
  },
  HeadingSemiBold18: {
    fontType: types.FontType.HEADING,
    fontSize: 18,
    fontWeight: types.FontWeights.SEMI_BOLD
  },
  HeadingSemiBold16: {
    fontType: types.FontType.HEADING,
    fontSize: 16,
    fontWeight: types.FontWeights.SEMI_BOLD
  },
  HeadingSemiBold14: {
    fontType: types.FontType.HEADING,
    fontSize: 14,
    fontWeight: types.FontWeights.SEMI_BOLD
  },
  HeadingSemiBold13: {
    fontType: types.FontType.HEADING,
    fontSize: 13,
    fontWeight: types.FontWeights.SEMI_BOLD
  },
  HeadingSemiBold12: {
    fontType: types.FontType.HEADING,
    fontSize: 12,
    fontWeight: types.FontWeights.SEMI_BOLD
  },
  HeadingSemiBold10: {
    fontType: types.FontType.HEADING,
    fontSize: 10,
    fontWeight: types.FontWeights.SEMI_BOLD
  },
  BodyMedium16: {
    fontType: types.FontType.BODY,
    fontSize: 16,
    fontWeight: types.FontWeights.MEDIUM
  },
  BodyMedium15: {
    fontType: types.FontType.BODY,
    fontSize: 15,
    fontWeight: types.FontWeights.MEDIUM
  },
  BodyMedium14: {
    fontType: types.FontType.BODY,
    fontSize: 14,
    fontWeight: types.FontWeights.MEDIUM
  },
  BodyMedium13: {
    fontType: types.FontType.BODY,
    fontSize: 13,
    fontWeight: types.FontWeights.MEDIUM
  },
  BodyMedium12: {
    fontType: types.FontType.BODY,
    fontSize: 12,
    fontWeight: types.FontWeights.MEDIUM
  },
  BodyMedium11: {
    fontType: types.FontType.BODY,
    fontSize: 11,
    fontWeight: types.FontWeights.MEDIUM
  },
  BodyRegular16: {
    fontType: types.FontType.BODY,
    fontSize: 16,
    fontWeight: types.FontWeights.REGULAR
  },
  BodyRegular15: {
    fontType: types.FontType.BODY,
    fontSize: 15,
    fontWeight: types.FontWeights.REGULAR
  },
  BodyRegular14: {
    fontType: types.FontType.BODY,
    fontSize: 14,
    fontWeight: types.FontWeights.REGULAR
  },
  BodyRegular13: {
    fontType: types.FontType.BODY,
    fontSize: 13,
    fontWeight: types.FontWeights.REGULAR
  },
  BodyRegular12: {
    fontType: types.FontType.BODY,
    fontSize: 12,
    fontWeight: types.FontWeights.REGULAR
  },
  BodyRegular11: {
    fontType: types.FontType.BODY,
    fontSize: 11,
    fontWeight: types.FontWeights.REGULAR
  },
  CapsExtraBold12: {
    fontType: types.FontType.CAPS,
    fontSize: 12,
    fontWeight: types.FontWeights.EXTRA_BOLD
  },
  CapsExtraBold10: {
    fontType: types.FontType.CAPS,
    fontSize: 10,
    fontWeight: types.FontWeights.EXTRA_BOLD
  },
  CapsBold12: {
    fontType: types.FontType.CAPS,
    fontSize: 12,
    fontWeight: types.FontWeights.BOLD
  },
  CapsBold10: {
    fontType: types.FontType.CAPS,
    fontSize: 10,
    fontWeight: types.FontWeights.BOLD
  },
  CapsBold8: {
    fontType: types.FontType.CAPS,
    fontSize: 8,
    fontWeight: types.FontWeights.BOLD
  },
  CapsSemiBold12: {
    fontType: types.FontType.CAPS,
    fontSize: 12,
    fontWeight: types.FontWeights.SEMI_BOLD
  },
  CapsSemiBold10: {
    fontType: types.FontType.CAPS,
    fontSize: 10,
    fontWeight: types.FontWeights.SEMI_BOLD
  },
  CirkaHeadingBold36: {
    fontType: types.FontType.SERIF_HEADING,
    fontSize: 36,
    fontWeight: types.FontWeights.BOLD
  },
  CirkaHeadingBold34: {
    fontType: types.FontType.SERIF_HEADING,
    fontSize: 34,
    fontWeight: types.FontWeights.BOLD
  },
  CirkaHeadingBold32: {
    fontType: types.FontType.SERIF_HEADING,
    fontSize: 32,
    fontWeight: types.FontWeights.BOLD
  },
  CirkaHeadingBold24: {
    fontType: types.FontType.SERIF_HEADING,
    fontSize: 24,
    fontWeight: types.FontWeights.BOLD
  },
  CirkaHeadingBold22: {
    fontType: types.FontType.SERIF_HEADING,
    fontSize: 22,
    fontWeight: types.FontWeights.BOLD
  },
  CirkaHeadingBold20: {
    fontType: types.FontType.SERIF_HEADING,
    fontSize: 20,
    fontWeight: types.FontWeights.BOLD
  },
  CirkaHeadingBold18: {
    fontType: types.FontType.SERIF_HEADING,
    fontSize: 18,
    fontWeight: types.FontWeights.BOLD
  }
};
var typographyGuide = {
  back: {
    heading: FontVariant.HeadingSemiBold14
  },
  header: {
    heading: FontVariant.CirkaHeadingBold24,
    description: FontVariant.BodyMedium14
  },
  containerCard: {
    heading: FontVariant.HeadingBold14,
    description: FontVariant.BodyMedium13
  },
  bottomSheet: {},
  tags: {
    containerText: FontVariant.CapsBold8,
    noContainerText: FontVariant.CapsBold10
  },
  searchBar: {
    input: FontVariant.HeadingSemiBold13
  },
  buttons: {
    big: FontVariant.HeadingBold14,
    medium: FontVariant.HeadingBold13,
    small: FontVariant.HeadingBold11
  },
  checkbox: {},
  toast: {
    heading: FontVariant.HeadingSemiBold13,
    description: FontVariant.BodyRegular11
  },
  inputFields: {
    label: FontVariant.CapsBold8,
    text: FontVariant.HeadingBold22
  }
};

exports.FontVariant = FontVariant;
exports.fontNameSpaces = fontNameSpaces;
exports.typographyGuide = typographyGuide;
