// @cred/neopop-web@1.0.0/lib-esm/primitives/buttons.js
import { objectSpread2 as _objectSpread2 } from '../_virtual/_rollupPluginBabelHelpers.js';
import { colorGuide } from './colors.js';
import { typographyGuide } from './typography.js';

var buttonTypography = typographyGuide.buttons;
var getButtonColors = function getButtonColors(mode, variant, kind) {
  var colors = mode === 'dark' ? colorGuide.darkComponents.buttons : colorGuide.lightComponents.buttons;
  var transparentEdge = kind === 'elevated' ? {
    top: 'transparent',
    left: 'transparent'
  } : {};
  return variant === 'primary' ? _objectSpread2(_objectSpread2({}, colors.primary), {}, {
    mode: mode,
    edgeColors: _objectSpread2(_objectSpread2({}, colors.primary.edgeColors), transparentEdge)
  }) : _objectSpread2(_objectSpread2({}, colors.secondary), {}, {
    mode: mode,
    edgeColors: _objectSpread2(_objectSpread2({}, colors.secondary.edgeColors), transparentEdge)
  });
};

var getSize = function getSize(number) {
  switch (number) {
    case '50':
      return 'big';

    case '40':
      return 'medium';

    case '30':
      return 'small';

    default:
      return 'medium';
  }
};

var sizeConfig = {
  big: {
    height: '50px',
    padding: '0px 30px',
    iconHeight: '20px'
  },
  medium: {
    height: '40px',
    padding: '0px 20px',
    iconHeight: '16px'
  },
  small: {
    height: '30px',
    padding: '0px 25px',
    iconHeight: '14px'
  }
};
var getSpacingConfig = function getSpacingConfig(size) {
  switch (size) {
    case 'big':
      return sizeConfig.big;

    case 'medium':
      return sizeConfig.medium;

    case 'small':
      return sizeConfig.small;

    default:
      return sizeConfig.medium;
  }
};
var getTextStyle = function getTextStyle(size) {
  switch (size) {
    case 'big':
      return buttonTypography.big;

    case 'medium':
      return buttonTypography.medium;

    case 'small':
      return buttonTypography.small;

    default:
      return buttonTypography.medium;
  }
};
/** ****************************************************************************************** */
// New Configs

var getButtonConfig = function getButtonConfig(name) {
  var colorMode = name.slice(1, 2) === 'l' ? 'light' : 'dark';
  var variant = name.slice(2, 3) === 's' ? 'secondary' : 'primary';
  var size = name.slice(3, 5);
  var kind = name.slice(5, 6) === 'p' ? 'elevated' : 'flat';
  var arrow = name.slice(6, 7);
  var config = {
    colorConfig: getButtonColors(colorMode, variant, kind),
    spacingConfig: getSpacingConfig(getSize(size)),
    colorMode: colorMode,
    size: size,
    kind: kind,
    variant: variant,
    showArrow: Boolean(Number(arrow))
  };
  return config;
};

export { getButtonColors, getButtonConfig, getSpacingConfig, getTextStyle };
