// @cred/neopop-web@1.0.0/lib-esm/components/Slider/index.js
import React__default from 'react';
import { SliderWrapper } from './styles.js';
import '../../primitives/index.js';
import { isEmpty } from '../../utils/index.js';
import { SliderConfig, ThumbConfig } from '../../primitives/slider.js';

var Slider = function Slider(props) {
  var min = props.min,
      max = props.max,
      step = props.step,
      defaultValue = props.defaultValue,
      value = props.value,
      onInput = props.onInput,
      onChange = props.onChange,
      style = props.style,
      sliderConfig = props.sliderConfig,
      thumbConfig = props.thumbConfig;
  var sliderUiConfig = isEmpty(sliderConfig) ? SliderConfig : sliderConfig;
  var thumbUiConfig = isEmpty(thumbConfig) ? ThumbConfig : thumbConfig;
  return /*#__PURE__*/React__default.createElement(SliderWrapper, {
    sliderConfig: sliderUiConfig,
    thumbConfig: thumbUiConfig
  }, /*#__PURE__*/React__default.createElement("input", {
    className: "slider",
    type: "range",
    min: min,
    max: max,
    step: step,
    defaultValue: defaultValue,
    value: value,
    onInput: onInput,
    onChange: onChange,
    style: style
  }));
};

export { Slider as default };
