// @cred/neopop-web@1.0.0/lib/components/Button/styles.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var styled = require('styled-components');
var index = require('../../primitives/index.js');
var types = require('./types.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var styled__default = /*#__PURE__*/_interopDefaultLegacy(styled);

var EDGEWIDTH = index.PlunkProps.WIDTH;
var SKEW_ANGLE = index.PlunkProps.ANGLE;

var shouldNotMoveRightEdge = function shouldNotMoveRightEdge(elevationDirection) {
  if (elevationDirection === types.NeoPopElevationDirections.lc || elevationDirection === types.NeoPopElevationDirections.rc || elevationDirection === types.NeoPopElevationDirections.tc || elevationDirection === types.NeoPopElevationDirections.tr || elevationDirection === types.NeoPopElevationDirections.tl) return true;
  return false;
};

var shouldNotMoveBottomEdge = function shouldNotMoveBottomEdge(elevationDirection) {
  if (elevationDirection === types.NeoPopElevationDirections.lc || elevationDirection === types.NeoPopElevationDirections.tc || elevationDirection === types.NeoPopElevationDirections.bc || elevationDirection === types.NeoPopElevationDirections.bl || elevationDirection === types.NeoPopElevationDirections.tl) return true;
  return false;
};

var ButtonWrapper = styled__default["default"].button.withConfig({
  displayName: "styles__ButtonWrapper",
  componentId: "sc-19j50wz-0"
})(["outline:none;margin:0;border:0;background:transparent;padding:0;display:", ";overflow:hidden;user-select:none;-webkit-tap-highlight-color:transparent;--webkit-user-select:none;cursor:pointer;overflow:hidden;position:relative;width:", ";height:100%;will-change:transform;", " .button-face{position:relative;display:flex;align-items:center;justify-content:center;text-align:center;background-color:", ";border:", ";padding:", ";z-index:2;height:", ";transition:transform 0.12s ease-in-out;-webkit-transition:transform 0.12s ease-in-out;-moz-transition:transform 0.12s ease-in-out;&.elevated{margin:0 ", "px ", "px 0;}.icon{height:", ";width:auto;margin-right:10px;}.arrow{margin-left:10px;}}.button-edge{position:absolute;transform-origin:0% 0%;transition:transform 0.12s ease-in-out;-webkit-transition:transform 0.12s ease-in-out;-moz-transition:transform 0.12s ease-in-out;&.right{height:calc(100% - ", "px);width:", "px;transform:translate3d(0px,0px,0px) skewY(45deg);right:0px;background-color:", ";top:0px;z-index:2;}&.bottom{width:calc(100% - ", "px);height:", "px;transform:translate3d(0px,0px,0px) skewX(45deg);bottom:0px;left:0px;background-color:", ";z-index:2;}&.top{width:calc(100% - ", "px);height:", "px;transform:skewX(45deg) scaleY(1);top:0px;left:0px;background-color:", ";&:after{content:'';height:3px;width:6px;background-color:", ";transform:skewX(-45deg) scaleY(1);position:absolute;right:-2px;}}&.left{height:calc(100% - ", "px);width:", "px;transform:skewY(45deg) scaleY(1);left:0px;top:0px;z-index:1;background-color:", ";&:after{content:'';height:6px;width:3px;background-color:", ";transform:skewY(-45deg) scaleY(1);position:absolute;bottom:-2px;right:0px;}}}&:active,&.active{.button-face{transform:translate3d(", "px,", "px,0);}.button-edge{&.bottom{transform:", ";}&.right{transform:", ";}}}", ""], function (_ref) {
  var fullWidth = _ref.fullWidth;
  return fullWidth ? 'block' : 'inline-flex';
}, function (_ref2) {
  var fullWidth = _ref2.fullWidth;
  return fullWidth ? '100%' : 'fit-content';
}, function (_ref3) {
  var disabled = _ref3.disabled;
  return disabled ? styled.css(["pointer-events:none;"]) : "\n\n    ";
}, function (_ref4) {
  var _colorConfig$backgrou;

  var colorConfig = _ref4.colorConfig;
  return (_colorConfig$backgrou = colorConfig === null || colorConfig === void 0 ? void 0 : colorConfig.backgroundColor) !== null && _colorConfig$backgrou !== void 0 ? _colorConfig$backgrou : 'transparent';
}, function (_ref5) {
  var colorConfig = _ref5.colorConfig;
  return colorConfig !== null && colorConfig !== void 0 && colorConfig.borderColor ? "1px solid ".concat(colorConfig === null || colorConfig === void 0 ? void 0 : colorConfig.borderColor) : 'none';
}, function (_ref6) {
  var _spacingConfig$paddin;

  var spacingConfig = _ref6.spacingConfig;
  return (_spacingConfig$paddin = spacingConfig === null || spacingConfig === void 0 ? void 0 : spacingConfig.padding) !== null && _spacingConfig$paddin !== void 0 ? _spacingConfig$paddin : '16px 20px';
}, function (_ref7) {
  var _spacingConfig$height;

  var spacingConfig = _ref7.spacingConfig;
  return (_spacingConfig$height = spacingConfig === null || spacingConfig === void 0 ? void 0 : spacingConfig.height) !== null && _spacingConfig$height !== void 0 ? _spacingConfig$height : 'auto';
}, EDGEWIDTH, EDGEWIDTH, function (_ref8) {
  var _spacingConfig$iconHe;

  var spacingConfig = _ref8.spacingConfig;
  return (_spacingConfig$iconHe = spacingConfig === null || spacingConfig === void 0 ? void 0 : spacingConfig.iconHeight) !== null && _spacingConfig$iconHe !== void 0 ? _spacingConfig$iconHe : '16px';
}, EDGEWIDTH, EDGEWIDTH, function (_ref9) {
  var _colorConfig$edgeColo, _colorConfig$edgeColo2;

  var colorConfig = _ref9.colorConfig;
  return (_colorConfig$edgeColo = colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$edgeColo2 = colorConfig.edgeColors) === null || _colorConfig$edgeColo2 === void 0 ? void 0 : _colorConfig$edgeColo2.right) !== null && _colorConfig$edgeColo !== void 0 ? _colorConfig$edgeColo : 'transparent';
}, EDGEWIDTH, EDGEWIDTH, function (_ref10) {
  var _colorConfig$edgeColo3, _colorConfig$edgeColo4;

  var colorConfig = _ref10.colorConfig;
  return (_colorConfig$edgeColo3 = colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$edgeColo4 = colorConfig.edgeColors) === null || _colorConfig$edgeColo4 === void 0 ? void 0 : _colorConfig$edgeColo4.bottom) !== null && _colorConfig$edgeColo3 !== void 0 ? _colorConfig$edgeColo3 : 'transparent';
}, EDGEWIDTH, EDGEWIDTH, function (_ref11) {
  var _colorConfig$edgeColo5, _colorConfig$edgeColo6;

  var colorConfig = _ref11.colorConfig;
  return (_colorConfig$edgeColo5 = colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$edgeColo6 = colorConfig.edgeColors) === null || _colorConfig$edgeColo6 === void 0 ? void 0 : _colorConfig$edgeColo6.top) !== null && _colorConfig$edgeColo5 !== void 0 ? _colorConfig$edgeColo5 : 'transparent';
}, function (_ref12) {
  var _colorConfig$edgeColo7, _colorConfig$edgeColo8;

  var colorConfig = _ref12.colorConfig;
  return (_colorConfig$edgeColo7 = colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$edgeColo8 = colorConfig.edgeColors) === null || _colorConfig$edgeColo8 === void 0 ? void 0 : _colorConfig$edgeColo8.top) !== null && _colorConfig$edgeColo7 !== void 0 ? _colorConfig$edgeColo7 : 'transparent';
}, EDGEWIDTH, EDGEWIDTH, function (_ref13) {
  var _colorConfig$edgeColo9, _colorConfig$edgeColo10;

  var colorConfig = _ref13.colorConfig;
  return (_colorConfig$edgeColo9 = colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$edgeColo10 = colorConfig.edgeColors) === null || _colorConfig$edgeColo10 === void 0 ? void 0 : _colorConfig$edgeColo10.left) !== null && _colorConfig$edgeColo9 !== void 0 ? _colorConfig$edgeColo9 : 'transparent';
}, function (_ref14) {
  var _colorConfig$edgeColo11, _colorConfig$edgeColo12;

  var colorConfig = _ref14.colorConfig;
  return (_colorConfig$edgeColo11 = colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$edgeColo12 = colorConfig.edgeColors) === null || _colorConfig$edgeColo12 === void 0 ? void 0 : _colorConfig$edgeColo12.left) !== null && _colorConfig$edgeColo11 !== void 0 ? _colorConfig$edgeColo11 : 'transparent';
}, EDGEWIDTH, EDGEWIDTH, function (_ref15) {
  var elevationDirection = _ref15.elevationDirection;
  return shouldNotMoveRightEdge(elevationDirection) ? "skewX(".concat(SKEW_ANGLE, "deg)") : "translate3d(".concat(EDGEWIDTH, "px, ").concat(EDGEWIDTH, "px, 0) skewX(").concat(SKEW_ANGLE, "deg)");
}, function (_ref16) {
  var elevationDirection = _ref16.elevationDirection;
  return shouldNotMoveBottomEdge(elevationDirection) ? "skewY(".concat(SKEW_ANGLE, "deg)") : "translate3d(".concat(EDGEWIDTH, "px, ").concat(EDGEWIDTH, "px, 0) skewY(").concat(SKEW_ANGLE, "deg)");
}, function (_ref17) {
  var disabled = _ref17.disabled;
  return disabled ? styled.css([".button-face{background-color:", ";border:", ";color:rgba(255,255,255,0.5);&:active{transform:none;}}.button-edge{&.right{background-color:", ";}&.bottom{background-color:", ";}&.top{background-color:", ";}&.left{background-color:", ";}&:active,&.active{&.bottom{transform:skewX(", "deg);}&.right{transform:skewY(", "deg);}}}"], function (_ref18) {
    var _colorConfig$disabled, _colorConfig$disabled2;

    var colorConfig = _ref18.colorConfig;
    return colorConfig !== null && colorConfig !== void 0 && (_colorConfig$disabled = colorConfig.disabledColors) !== null && _colorConfig$disabled !== void 0 && _colorConfig$disabled.backgroundColor ? colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$disabled2 = colorConfig.disabledColors) === null || _colorConfig$disabled2 === void 0 ? void 0 : _colorConfig$disabled2.backgroundColor : '';
  }, function (_ref19) {
    var _colorConfig$disabled3, _colorConfig$disabled4;

    var colorConfig = _ref19.colorConfig;
    return colorConfig !== null && colorConfig !== void 0 && (_colorConfig$disabled3 = colorConfig.disabledColors) !== null && _colorConfig$disabled3 !== void 0 && _colorConfig$disabled3.borderColor ? "1px solid ".concat(colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$disabled4 = colorConfig.disabledColors) === null || _colorConfig$disabled4 === void 0 ? void 0 : _colorConfig$disabled4.borderColor) : '';
  }, function (_ref20) {
    var _colorConfig$disabled5, _colorConfig$disabled6, _colorConfig$disabled7;

    var colorConfig = _ref20.colorConfig;
    return (_colorConfig$disabled5 = colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$disabled6 = colorConfig.disabledColors) === null || _colorConfig$disabled6 === void 0 ? void 0 : (_colorConfig$disabled7 = _colorConfig$disabled6.edgeColors) === null || _colorConfig$disabled7 === void 0 ? void 0 : _colorConfig$disabled7.right) !== null && _colorConfig$disabled5 !== void 0 ? _colorConfig$disabled5 : 'transparent';
  }, function (_ref21) {
    var _colorConfig$disabled8, _colorConfig$disabled9, _colorConfig$disabled10;

    var colorConfig = _ref21.colorConfig;
    return (_colorConfig$disabled8 = colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$disabled9 = colorConfig.disabledColors) === null || _colorConfig$disabled9 === void 0 ? void 0 : (_colorConfig$disabled10 = _colorConfig$disabled9.edgeColors) === null || _colorConfig$disabled10 === void 0 ? void 0 : _colorConfig$disabled10.bottom) !== null && _colorConfig$disabled8 !== void 0 ? _colorConfig$disabled8 : 'transparent';
  }, function (_ref22) {
    var _colorConfig$disabled11, _colorConfig$disabled12, _colorConfig$disabled13;

    var colorConfig = _ref22.colorConfig;
    return (_colorConfig$disabled11 = colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$disabled12 = colorConfig.disabledColors) === null || _colorConfig$disabled12 === void 0 ? void 0 : (_colorConfig$disabled13 = _colorConfig$disabled12.edgeColors) === null || _colorConfig$disabled13 === void 0 ? void 0 : _colorConfig$disabled13.top) !== null && _colorConfig$disabled11 !== void 0 ? _colorConfig$disabled11 : 'transparent';
  }, function (_ref23) {
    var _colorConfig$disabled14, _colorConfig$disabled15, _colorConfig$disabled16;

    var colorConfig = _ref23.colorConfig;
    return (_colorConfig$disabled14 = colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$disabled15 = colorConfig.disabledColors) === null || _colorConfig$disabled15 === void 0 ? void 0 : (_colorConfig$disabled16 = _colorConfig$disabled15.edgeColors) === null || _colorConfig$disabled16 === void 0 ? void 0 : _colorConfig$disabled16.left) !== null && _colorConfig$disabled14 !== void 0 ? _colorConfig$disabled14 : 'transparent';
  }, SKEW_ANGLE, SKEW_ANGLE) : "\n\n        ";
});
var LinkButtonContainer = styled__default["default"].button.withConfig({
  displayName: "styles__LinkButtonContainer",
  componentId: "sc-19j50wz-1"
})(["outline:none;background-color:transparent;margin:0;padding:0;border:0;display:inline-flex;user-select:none;-webkit-tap-highlight-color:transparent;--webkit-user-select:none;cursor:pointer;color:", ";border-bottom:", ";", ""], function (_ref24) {
  var color = _ref24.color;
  return color !== null && color !== void 0 ? color : '';
}, function (_ref25) {
  var color = _ref25.color;
  return color ? "1px solid ".concat(color) : '';
}, function (_ref26) {
  var disabled = _ref26.disabled;
  return disabled ? styled.css(["pointer-events:none;"]) : "\n\n        ";
});

exports.ButtonWrapper = ButtonWrapper;
exports.LinkButtonContainer = LinkButtonContainer;
