// @cred/neopop-web@1.0.0/lib/components/Checkbox/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../_virtual/_rollupPluginBabelHelpers.js');
var React = require('react');
var withCheckHOC = require('../Helpers/withCheckHOC.js');
var styles = require('./styles.js');
require('../../primitives/index.js');
var index$1 = require('../Helpers/index.js');
var index$2 = require('../../utils/index.js');
var colors = require('../../primitives/colors.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var _excluded = ["isChecked", "handleChange", "colorConfig", "colorMode", "children"];

var Checkbox = function Checkbox(props) {
  var isChecked = props.isChecked,
      handleChange = props.handleChange,
      colorConfig = props.colorConfig,
      colorMode = props.colorMode,
      children = props.children,
      propsToFwd = _rollupPluginBabelHelpers.objectWithoutProperties(props, _excluded);

  var defaultColorConfig = colors.colorGuide[colorMode === 'light' ? 'lightComponents' : 'darkComponents'].checkbox;
  var checkboxColorConfig = index$2.isEmpty(colorConfig) ? defaultColorConfig : colorConfig !== null && colorConfig !== void 0 ? colorConfig : defaultColorConfig;
  return /*#__PURE__*/React__default["default"].createElement(styles.CheckboxLabel, null, /*#__PURE__*/React__default["default"].createElement("input", _rollupPluginBabelHelpers["extends"]({}, propsToFwd, {
    type: "checkbox",
    checked: isChecked,
    onChange: handleChange
  })), /*#__PURE__*/React__default["default"].createElement(index$1.Row, null, /*#__PURE__*/React__default["default"].createElement(styles.CheckmarkWrapper, {
    colorConfig: checkboxColorConfig,
    hasChildren: children ? true : false,
    checked: isChecked
  }, /*#__PURE__*/React__default["default"].createElement("span", {
    className: "checkmark"
  }, /*#__PURE__*/React__default["default"].createElement("svg", {
    className: "tick-svg",
    width: "10",
    height: "8",
    viewBox: "0 0 11 8",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/React__default["default"].createElement("path", {
    d: "M1.45456 3.99973L4.18183 6.72701L10.5455 1.27246",
    stroke: checkboxColorConfig.background,
    strokeWidth: "1.36364"
  })))), children ? /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement(index$1.VerticalSpacer, {
    n: 3
  }), children) : null));
};

var index = withCheckHOC["default"](Checkbox);

exports["default"] = index;
