// @cred/neopop-web@1.0.0/lib-esm/components/Toast/index.js
import { createClass as _createClass, objectSpread2 as _objectSpread2, classCallCheck as _classCallCheck, defineProperty as _defineProperty } from '../../_virtual/_rollupPluginBabelHelpers.js';
import React__default, { useRef, useEffect } from 'react';
import { ToastPortal } from './components/ToastPortal.js';

var GlobalStore = /*#__PURE__*/function () {
  function GlobalStore() {
    _classCallCheck(this, GlobalStore);

    _defineProperty(this, "func", void 0);

    this.func = [];
  }

  _createClass(GlobalStore, [{
    key: "on",
    value: function on(listener) {
      this.func.push(listener);
    }
  }, {
    key: "emit",
    value: function emit(options) {
      this.func.forEach(function (l) {
        return l(options);
      });
    }
  }]);

  return GlobalStore;
}();

var globalStore = new GlobalStore();
var showToast = function showToast(content, options) {
  globalStore.emit(_objectSpread2(_objectSpread2({}, options), {}, {
    content: content
  }));
};
var ToastContainer = function ToastContainer() {
  var toastRef = useRef();
  useEffect(function () {
    var listener = function listener(options) {
      var _toastRef$current;

      (_toastRef$current = toastRef.current) === null || _toastRef$current === void 0 ? void 0 : _toastRef$current.addToast(options);
    };

    globalStore.on(listener);
  }, []);
  return /*#__PURE__*/React__default.createElement(ToastPortal, {
    ref: toastRef
  });
};

export { ToastContainer, showToast };
