// @cred/neopop-web@1.0.0/lib/primitives/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var buttons = require('./buttons.js');
var scoremeter = require('./scoremeter.js');
var colors = require('./colors.js');
var typography = require('./typography.js');
var slider = require('./slider.js');
var opacity = require('./opacity.js');

exports.PlunkProps = void 0;

(function (PlunkProps) {
  PlunkProps[PlunkProps["WIDTH"] = 3] = "WIDTH";
  PlunkProps[PlunkProps["ANGLE"] = 45] = "ANGLE";
})(exports.PlunkProps || (exports.PlunkProps = {}));

exports.getButtonConfig = buttons.getButtonConfig;
exports.getScoreMeterConfig = scoremeter.getScoreMeterConfig;
Object.defineProperty(exports, 'colorGuide', {
    enumerable: true,
    get: function () { return colors.colorGuide; }
});
Object.defineProperty(exports, 'colorPalette', {
    enumerable: true,
    get: function () { return colors.colorPalette; }
});
exports.extend = colors.extend;
Object.defineProperty(exports, 'mainColors', {
    enumerable: true,
    get: function () { return colors.mainColors; }
});
exports.FontVariant = typography.FontVariant;
exports.fontNameSpaces = typography.fontNameSpaces;
exports.typographyGuide = typography.typographyGuide;
exports.SliderConfig = slider.SliderConfig;
exports.ThumbConfig = slider.ThumbConfig;
Object.defineProperty(exports, 'FontOpacity', {
    enumerable: true,
    get: function () { return opacity.FontOpacity; }
});
