// @cred/neopop-web@1.0.0/lib-esm/components/Toggle/index.js
import React__default, { useRef } from 'react';
import withCheckHOC from '../Helpers/withCheckHOC.js';
import { ToggleBtn } from './styles.js';
import '../../primitives/index.js';
import { isEmpty } from '../../utils/index.js';
import { colorGuide } from '../../primitives/colors.js';

var Toggle = function Toggle(props) {
  var isChecked = props.isChecked,
      handleChange = props.handleChange,
      colorConfig = props.colorConfig,
      colorMode = props.colorMode;
  var defaultColorConfig = colorGuide[colorMode === 'light' ? 'lightComponents' : 'darkComponents'].toggle;
  var toggleRef = useRef(null);
  return /*#__PURE__*/React__default.createElement(ToggleBtn, {
    isChecked: isChecked,
    colorConfig: isEmpty(colorConfig) ? defaultColorConfig : colorConfig !== null && colorConfig !== void 0 ? colorConfig : defaultColorConfig,
    onClick: function onClick() {
      var _toggleRef$current;

      return toggleRef === null || toggleRef === void 0 ? void 0 : (_toggleRef$current = toggleRef.current) === null || _toggleRef$current === void 0 ? void 0 : _toggleRef$current.click();
    }
  }, /*#__PURE__*/React__default.createElement("input", {
    type: "checkbox",
    ref: toggleRef,
    checked: isChecked,
    onChange: handleChange
  }), /*#__PURE__*/React__default.createElement("div", {
    className: "mark"
  }));
};

var index = withCheckHOC(Toggle);

export { index as default };
