// @cred/neopop-web@1.0.0/lib/components/Dropdown/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../_virtual/_rollupPluginBabelHelpers.js');
var React = require('react');
var index$1 = require('../Helpers/index.js');
var index = require('../Typography/index.js');
require('../../primitives/index.js');
require('../Button/types.js');
require('../Typography/types.js');
var types = require('../Helpers/types.js');
var styles = require('./styles.js');
var index$2 = require('../../utils/index.js');
var colors = require('../../primitives/colors.js');
var typography = require('../../primitives/typography.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var Dropdown = function Dropdown(props) {
  var _dropdownColorConfig$, _dropdownColorConfig$2;

  var onClick = props.onClick,
      label = props.label,
      colorConfig = props.colorConfig,
      colorMode = props.colorMode;
  var defaultColorConfig = colors.colorGuide[colorMode === 'light' ? 'lightComponents' : 'darkComponents'].dropdown[label ? 'withLabel' : 'withoutLabel'];
  var dropdownColorConfig = index$2.isEmpty(colorConfig) ? defaultColorConfig : colorConfig !== null && colorConfig !== void 0 ? colorConfig : defaultColorConfig;
  return /*#__PURE__*/React__default["default"].createElement(styles.DropdownContainer, _rollupPluginBabelHelpers["extends"]({
    className: label ? '' : 'no-label'
  }, dropdownColorConfig, {
    onClick: onClick
  }), label ? /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement(index["default"], _rollupPluginBabelHelpers["extends"]({}, typography.fontNameSpaces.th13sb, {
    color: (_dropdownColorConfig$ = dropdownColorConfig === null || dropdownColorConfig === void 0 ? void 0 : dropdownColorConfig.text) !== null && _dropdownColorConfig$ !== void 0 ? _dropdownColorConfig$ : 'rgba(255,255,255,0.5)'
  }), label), /*#__PURE__*/React__default["default"].createElement(index$1.VerticalSpacer, {
    n: 2
  })) : null, /*#__PURE__*/React__default["default"].createElement(index$1.Chevron, {
    direction: types.Directions.SOUTH,
    color: (_dropdownColorConfig$2 = dropdownColorConfig === null || dropdownColorConfig === void 0 ? void 0 : dropdownColorConfig.chevron) !== null && _dropdownColorConfig$2 !== void 0 ? _dropdownColorConfig$2 : 'rgba(255,255,255,0.5)'
  }));
};

exports["default"] = Dropdown;
