// @cred/neopop-web@1.0.0/lib/components/Button/components/FlatButton.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../../_virtual/_rollupPluginBabelHelpers.js');
require('../../../primitives/index.js');
var buttons = require('../../../primitives/buttons.js');
var React = require('react');
var index$1 = require('../../Helpers/index.js');
var index = require('../../Typography/index.js');
var styles = require('../styles.js');
var typography = require('../../../primitives/typography.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var _excluded = ["variant", "colorMode", "kind", "size", "textStyle", "colorConfig", "spacingConfig", "children", "onTouchStart"];

var FlatButton = function FlatButton(props) {
  var _ref, _colors$disabledColor;

  var variant = props.variant,
      colorMode = props.colorMode,
      kind = props.kind,
      size = props.size,
      textStyle = props.textStyle,
      colorConfig = props.colorConfig,
      spacingConfig = props.spacingConfig,
      children = props.children,
      _props$onTouchStart = props.onTouchStart,
      onTouchStart = _props$onTouchStart === void 0 ? function () {} : _props$onTouchStart,
      propsToFwd = _rollupPluginBabelHelpers.objectWithoutProperties(props, _excluded);

  var spacing = spacingConfig !== null && spacingConfig !== void 0 ? spacingConfig : buttons.getSpacingConfig(size !== null && size !== void 0 ? size : 'medium');
  var colors = colorConfig !== null && colorConfig !== void 0 ? colorConfig : buttons.getButtonColors(colorMode !== null && colorMode !== void 0 ? colorMode : 'dark', variant !== null && variant !== void 0 ? variant : 'primary', kind !== null && kind !== void 0 ? kind : 'flat');
  var customTextStyle = (_ref = textStyle !== null && textStyle !== void 0 ? textStyle : buttons.getTextStyle(size !== null && size !== void 0 ? size : 'medium')) !== null && _ref !== void 0 ? _ref : typography.fontNameSpaces.th14b;
  var textColor = propsToFwd.disabled ? colors === null || colors === void 0 ? void 0 : (_colors$disabledColor = colors.disabledColors) === null || _colors$disabledColor === void 0 ? void 0 : _colors$disabledColor.color : colors === null || colors === void 0 ? void 0 : colors.color;
  return /*#__PURE__*/React__default["default"].createElement(styles.ButtonWrapper, _rollupPluginBabelHelpers["extends"]({
    colorConfig: colors,
    spacingConfig: spacing,
    textStyle: customTextStyle,
    color: textColor,
    onTouchStart: onTouchStart
  }, propsToFwd), /*#__PURE__*/React__default["default"].createElement("div", {
    className: "button-face"
  }, propsToFwd.icon ? /*#__PURE__*/React__default["default"].createElement("img", {
    src: propsToFwd.icon,
    className: "icon",
    alt: "icon"
  }) : null, /*#__PURE__*/React__default["default"].createElement(index["default"], {
    fontSize: customTextStyle === null || customTextStyle === void 0 ? void 0 : customTextStyle.fontSize,
    fontType: customTextStyle === null || customTextStyle === void 0 ? void 0 : customTextStyle.fontType,
    fontWeight: customTextStyle === null || customTextStyle === void 0 ? void 0 : customTextStyle.fontWeight,
    color: textColor
  }, children), propsToFwd.showArrow ? /*#__PURE__*/React__default["default"].createElement(index$1.Pointer, {
    style: {
      marginLeft: '10px'
    },
    color: textColor
  }) : null), /*#__PURE__*/React__default["default"].createElement("div", {
    className: "button-edge left"
  }), /*#__PURE__*/React__default["default"].createElement("div", {
    className: "button-edge top"
  }));
};

exports["default"] = FlatButton;
