// @cred/neopop-web@1.0.0/lib-esm/components/Checkbox/index.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../_virtual/_rollupPluginBabelHelpers.js';
import React__default from 'react';
import withCheckHOC from '../Helpers/withCheckHOC.js';
import { CheckboxLabel, CheckmarkWrapper } from './styles.js';
import '../../primitives/index.js';
import { Row, VerticalSpacer } from '../Helpers/index.js';
import { isEmpty } from '../../utils/index.js';
import { colorGuide } from '../../primitives/colors.js';

var _excluded = ["isChecked", "handleChange", "colorConfig", "colorMode", "children"];

var Checkbox = function Checkbox(props) {
  var isChecked = props.isChecked,
      handleChange = props.handleChange,
      colorConfig = props.colorConfig,
      colorMode = props.colorMode,
      children = props.children,
      propsToFwd = _objectWithoutProperties(props, _excluded);

  var defaultColorConfig = colorGuide[colorMode === 'light' ? 'lightComponents' : 'darkComponents'].checkbox;
  var checkboxColorConfig = isEmpty(colorConfig) ? defaultColorConfig : colorConfig !== null && colorConfig !== void 0 ? colorConfig : defaultColorConfig;
  return /*#__PURE__*/React__default.createElement(CheckboxLabel, null, /*#__PURE__*/React__default.createElement("input", _extends({}, propsToFwd, {
    type: "checkbox",
    checked: isChecked,
    onChange: handleChange
  })), /*#__PURE__*/React__default.createElement(Row, null, /*#__PURE__*/React__default.createElement(CheckmarkWrapper, {
    colorConfig: checkboxColorConfig,
    hasChildren: children ? true : false,
    checked: isChecked
  }, /*#__PURE__*/React__default.createElement("span", {
    className: "checkmark"
  }, /*#__PURE__*/React__default.createElement("svg", {
    className: "tick-svg",
    width: "10",
    height: "8",
    viewBox: "0 0 11 8",
    fill: "none",
    xmlns: "http://www.w3.org/2000/svg"
  }, /*#__PURE__*/React__default.createElement("path", {
    d: "M1.45456 3.99973L4.18183 6.72701L10.5455 1.27246",
    stroke: checkboxColorConfig.background,
    strokeWidth: "1.36364"
  })))), children ? /*#__PURE__*/React__default.createElement(React__default.Fragment, null, /*#__PURE__*/React__default.createElement(VerticalSpacer, {
    n: 3
  }), children) : null));
};

var index = withCheckHOC(Checkbox);

export { index as default };
