// @cred/neopop-web@1.0.0/lib/components/Scoremeter/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../_virtual/_rollupPluginBabelHelpers.js');
var React = require('react');
var index$1 = require('../Typography/index.js');
require('../../primitives/index.js');
var styles = require('./styles.js');
var types = require('./types.js');
var index = require('../Helpers/index.js');
var colors = require('../../primitives/colors.js');
var typography = require('../../primitives/typography.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var getScoremeterReading = function getScoremeterReading(reading, lowerLimit, upperLimit) {
  return (75 - 0) / (upperLimit - lowerLimit) * (reading - upperLimit) + 75;
};

var getDotRotateDegree = function getDotRotateDegree(reading, lowerLimit, upperLimit) {
  return 2.7 * ((reading - lowerLimit) * 100 / (upperLimit - lowerLimit)) + 47;
};

var ScoreMeter = function ScoreMeter(props) {
  var _colorConfig$indicato, _colorConfig$indicato2, _colorConfig$indicato3, _colorConfig$indicato4;

  var reading = props.reading,
      oldReading = props.oldReading,
      _props$colorMode = props.colorMode,
      colorMode = _props$colorMode === void 0 ? types.DEFAULT_MODE : _props$colorMode,
      _props$type = props.type,
      type = _props$type === void 0 ? types.DEFAULT_TYPE : _props$type,
      _props$colorConfig = props.colorConfig,
      colorConfig = _props$colorConfig === void 0 ? colors.colorGuide[colorMode == 'light' ? 'lightComponents' : 'darkComponents'].scoremeter : _props$colorConfig,
      scoreDesc = props.scoreDesc,
      showIndicators = props.showIndicators,
      _props$lowerLimit = props.lowerLimit,
      lowerLimit = _props$lowerLimit === void 0 ? 300 : _props$lowerLimit,
      _props$upperLimit = props.upperLimit,
      upperLimit = _props$upperLimit === void 0 ? 900 : _props$upperLimit,
      _props$showLegends = props.showLegends,
      showLegends = _props$showLegends === void 0 ? true : _props$showLegends;
  var sanitizedReading = reading < 0 ? types.LOWER_THRESHOLD : reading;
  var sanitizedOldReading = oldReading ? oldReading < 0 ? types.LOWER_THRESHOLD : oldReading : types.LOWER_THRESHOLD;

  var _React$useState = React__default["default"].useState(sanitizedReading),
      _React$useState2 = _rollupPluginBabelHelpers.slicedToArray(_React$useState, 2),
      score = _React$useState2[0],
      setScore = _React$useState2[1];

  var _React$useState3 = React__default["default"].useState(sanitizedOldReading),
      _React$useState4 = _rollupPluginBabelHelpers.slicedToArray(_React$useState3, 1),
      oldScore = _React$useState4[0];

  var _React$useState5 = React__default["default"].useState(false),
      _React$useState6 = _rollupPluginBabelHelpers.slicedToArray(_React$useState5, 2),
      animate = _React$useState6[0],
      setAnimate = _React$useState6[1];

  var _React$useState7 = React__default["default"].useState(false),
      _React$useState8 = _rollupPluginBabelHelpers.slicedToArray(_React$useState7, 2),
      scoreUpdate = _React$useState8[0],
      setScoreUpdate = _React$useState8[1];

  var differenceScore = sanitizedOldReading ? sanitizedReading - sanitizedOldReading : sanitizedReading;
  var differential = differenceScore < 0 ? -1 : 1;
  React__default["default"].useEffect(function () {
    if (sanitizedOldReading && sanitizedReading !== sanitizedOldReading) {
      setScore(sanitizedOldReading);
      setAnimate(function (prev) {
        return !prev;
      });
    }
  }, []);
  React__default["default"].useEffect(function () {
    if (score !== sanitizedReading) {
      setTimeout(function () {
        setScoreUpdate(true);
      }, 500);
    }
  }, [animate]);
  React__default["default"].useEffect(function () {
    if (Math.floor(sanitizedReading) !== Math.floor(score)) {
      setTimeout(function () {
        setScore(function (prevScore) {
          return prevScore + differential;
        });
        setScoreUpdate(function (prev) {
          return !prev;
        });
      }, 20);
    }
  }, [scoreUpdate]);
  var scoreReading = getScoremeterReading(score, lowerLimit, upperLimit);
  var dotRotateDegree = getDotRotateDegree(score, lowerLimit, upperLimit);
  return /*#__PURE__*/React__default["default"].createElement(index.Row, null, /*#__PURE__*/React__default["default"].createElement(styles.MeterContainer, null, /*#__PURE__*/React__default["default"].createElement(styles.MeterWrapper, {
    colorConfig: colorConfig,
    type: type,
    showIndicators: showIndicators
  }, /*#__PURE__*/React__default["default"].createElement("div", {
    className: "dot",
    style: {
      transform: "rotateZ(".concat(dotRotateDegree, "deg) translateY(80px)")
    }
  }), /*#__PURE__*/React__default["default"].createElement("div", {
    className: "meter"
  }, /*#__PURE__*/React__default["default"].createElement(styles.MeterSVG, {
    height: "140",
    width: "140",
    viewBox: "0 0 20 20"
  }, /*#__PURE__*/React__default["default"].createElement("circle", {
    r: "10",
    cx: "10",
    cy: "10",
    fill: "none"
  }), score > oldScore ? /*#__PURE__*/React__default["default"].createElement("circle", {
    r: "5",
    cx: "10",
    cy: "10",
    fill: "none",
    stroke: colorConfig === null || colorConfig === void 0 ? void 0 : colorConfig.meterStrokeColor[type !== null && type !== void 0 ? type : 'excellent'],
    strokeWidth: "10",
    strokeDasharray: "".concat(scoreReading * 31.42 / 100, " 31.42")
  }) : /*#__PURE__*/React__default["default"].createElement("circle", {
    r: "5",
    cx: "10",
    cy: "10",
    fill: "none",
    stroke: colorConfig === null || colorConfig === void 0 ? void 0 : colorConfig.meterStrokeColor[type !== null && type !== void 0 ? type : 'excellent'],
    strokeWidth: "10",
    strokeDasharray: "".concat((scoreReading + 0.75) * 31.42 / 100, " 31.42")
  }))), /*#__PURE__*/React__default["default"].createElement(styles.ScoreContainer, {
    colorConfig: colorConfig,
    type: type,
    showIndicators: showIndicators
  }, /*#__PURE__*/React__default["default"].createElement("div", {
    className: "score"
  }, /*#__PURE__*/React__default["default"].createElement(index$1["default"], _rollupPluginBabelHelpers["extends"]({}, typography.fontNameSpaces.th28eb, {
    style: {
      color: 'inherit'
    }
  }), sanitizedOldReading ? score : sanitizedReading), scoreDesc ? /*#__PURE__*/React__default["default"].createElement(index$1["default"], _rollupPluginBabelHelpers["extends"]({}, typography.fontNameSpaces.tc8b, {
    color: colorConfig === null || colorConfig === void 0 ? void 0 : colorConfig.meterStrokeColor[type],
    style: {
      marginTop: '5px'
    }
  }), scoreDesc) : null, showIndicators ? /*#__PURE__*/React__default["default"].createElement("div", null, /*#__PURE__*/React__default["default"].createElement("svg", {
    width: "15",
    height: "24",
    viewBox: "0 0 14 10"
  }, /*#__PURE__*/React__default["default"].createElement("path", {
    d: "M7.368 0L0 10h14z",
    fill: score > oldScore ? colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$indicato = colorConfig.indicatorColors) === null || _colorConfig$indicato === void 0 ? void 0 : _colorConfig$indicato.increment : colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$indicato2 = colorConfig.indicatorColors) === null || _colorConfig$indicato2 === void 0 ? void 0 : _colorConfig$indicato2.neutral,
    fillRule: "evenodd"
  })), /*#__PURE__*/React__default["default"].createElement("svg", {
    width: "15",
    height: "24",
    viewBox: "0 0 14 10"
  }, /*#__PURE__*/React__default["default"].createElement("path", {
    d: "M7.368 10L0 0h14z",
    fill: score < oldScore ? colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$indicato3 = colorConfig.indicatorColors) === null || _colorConfig$indicato3 === void 0 ? void 0 : _colorConfig$indicato3.decrement : colorConfig === null || colorConfig === void 0 ? void 0 : (_colorConfig$indicato4 = colorConfig.indicatorColors) === null || _colorConfig$indicato4 === void 0 ? void 0 : _colorConfig$indicato4.neutral,
    fillRule: "evenodd"
  }))) : ''))), showLegends ? /*#__PURE__*/React__default["default"].createElement(React__default["default"].Fragment, null, /*#__PURE__*/React__default["default"].createElement(styles.Legend, null, /*#__PURE__*/React__default["default"].createElement("div", {
    className: "dot"
  }, "\xB7"), /*#__PURE__*/React__default["default"].createElement(index$1["default"], _rollupPluginBabelHelpers["extends"]({}, typography.fontNameSpaces.tb14sb, {
    color: "#4b4f52"
  }), lowerLimit)), /*#__PURE__*/React__default["default"].createElement(styles.Legend, {
    right: true
  }, /*#__PURE__*/React__default["default"].createElement("div", {
    className: "dot"
  }, "\xB7"), /*#__PURE__*/React__default["default"].createElement(index$1["default"], _rollupPluginBabelHelpers["extends"]({}, typography.fontNameSpaces.tb14sb, {
    color: "#4b4f52"
  }), upperLimit))) : null));
};

exports["default"] = ScoreMeter;
