// @cred/neopop-web@1.0.0/lib/components/Scoremeter/styles.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var styled = require('styled-components');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var styled__default = /*#__PURE__*/_interopDefaultLegacy(styled);

var MeterContainer = styled__default["default"].div.withConfig({
  displayName: "styles__MeterContainer",
  componentId: "sc-1j4x4ch-0"
})(["position:relative;height:185px;"]);
var MeterWrapper = styled__default["default"].div.withConfig({
  displayName: "styles__MeterWrapper",
  componentId: "sc-1j4x4ch-1"
})(["position:relative;width:160px;height:160px;border-radius:50%;background:", ";display:flex;justify-content:center;align-items:center;border:", ";&.absoulte-center{left:50%;top:50%;-webkit-transform:translate(-50%,-50%);transform:translate(-50%,-50%);}.dot{width:6px;height:6px;position:absolute;background:", ";border-radius:50%;z-index:9;transform:rotateZ(45deg) translateY(80px);transition:transform 0.2s ease;}.meter{transform:rotate(135deg);background:", ";border-radius:50%;}"], function (_ref) {
  var _colorConfig$meterStr;

  var colorConfig = _ref.colorConfig;
  return (_colorConfig$meterStr = colorConfig === null || colorConfig === void 0 ? void 0 : colorConfig.meterStrokeBackground) !== null && _colorConfig$meterStr !== void 0 ? _colorConfig$meterStr : 'transparent';
}, function (_ref2) {
  var colorConfig = _ref2.colorConfig;
  return "1px solid ".concat(colorConfig === null || colorConfig === void 0 ? void 0 : colorConfig.meterBorderColor);
}, function (_ref3) {
  var colorConfig = _ref3.colorConfig,
      type = _ref3.type;
  return colorConfig !== null && colorConfig !== void 0 && colorConfig.dotColor ? colorConfig.dotColor : colorConfig === null || colorConfig === void 0 ? void 0 : colorConfig.meterStrokeColor[type !== null && type !== void 0 ? type : 'excellent'];
}, function (_ref4) {
  var colorConfig = _ref4.colorConfig;
  return colorConfig === null || colorConfig === void 0 ? void 0 : colorConfig.meterStrokeBackground;
});
var ScoreContainer = styled__default["default"].div.withConfig({
  displayName: "styles__ScoreContainer",
  componentId: "sc-1j4x4ch-2"
})(["width:115px;height:115px;border-radius:50%;background:", ";position:absolute;display:flex;flex-direction:column;justify-content:center;align-items:center;left:50%;top:50%;-webkit-transform:translate(-50%,-50%);transform:translate(-50%,-50%);.score{all:inherit;background:", ";color:", ";padding-top:", ";width:95px;height:95px;letter-spacing:1px;overflow:hidden;}"], function (_ref5) {
  var colorConfig = _ref5.colorConfig;
  return colorConfig !== null && colorConfig !== void 0 && colorConfig.scoreContainerBorder ? colorConfig === null || colorConfig === void 0 ? void 0 : colorConfig.scoreContainerBorder : 'transparent';
}, function (_ref6) {
  var colorConfig = _ref6.colorConfig;
  return colorConfig === null || colorConfig === void 0 ? void 0 : colorConfig.scoreContainerBackground;
}, function (_ref7) {
  var colorConfig = _ref7.colorConfig;
  return colorConfig === null || colorConfig === void 0 ? void 0 : colorConfig.scoreColor;
}, function (_ref8) {
  var showIndicators = _ref8.showIndicators;
  return showIndicators ? '6%' : '0';
});
var MeterSVG = styled__default["default"].svg.withConfig({
  displayName: "styles__MeterSVG",
  componentId: "sc-1j4x4ch-3"
})(["display:block;& .path{transition:stroke-dasharray 100ms ease;}& .path1{transition:stroke-dasharray 100ms ease;}"]);
var Legend = styled__default["default"].div.withConfig({
  displayName: "styles__Legend",
  componentId: "sc-1j4x4ch-4"
})(["position:absolute;bottom:0;color:#4b4f52;", ";.dot{text-align:center;color:#434547;}"], function (props) {
  return props.right ? styled.css(["right:0;"]) : styled.css(["left:0;"]);
});

exports.Legend = Legend;
exports.MeterContainer = MeterContainer;
exports.MeterSVG = MeterSVG;
exports.MeterWrapper = MeterWrapper;
exports.ScoreContainer = ScoreContainer;
