// @cred/neopop-web@1.0.0/lib/components/Toggle/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var React = require('react');
var withCheckHOC = require('../Helpers/withCheckHOC.js');
var styles = require('./styles.js');
require('../../primitives/index.js');
var index$1 = require('../../utils/index.js');
var colors = require('../../primitives/colors.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var Toggle = function Toggle(props) {
  var isChecked = props.isChecked,
      handleChange = props.handleChange,
      colorConfig = props.colorConfig,
      colorMode = props.colorMode;
  var defaultColorConfig = colors.colorGuide[colorMode === 'light' ? 'lightComponents' : 'darkComponents'].toggle;
  var toggleRef = React.useRef(null);
  return /*#__PURE__*/React__default["default"].createElement(styles.ToggleBtn, {
    isChecked: isChecked,
    colorConfig: index$1.isEmpty(colorConfig) ? defaultColorConfig : colorConfig !== null && colorConfig !== void 0 ? colorConfig : defaultColorConfig,
    onClick: function onClick() {
      var _toggleRef$current;

      return toggleRef === null || toggleRef === void 0 ? void 0 : (_toggleRef$current = toggleRef.current) === null || _toggleRef$current === void 0 ? void 0 : _toggleRef$current.click();
    }
  }, /*#__PURE__*/React__default["default"].createElement("input", {
    type: "checkbox",
    ref: toggleRef,
    checked: isChecked,
    onChange: handleChange
  }), /*#__PURE__*/React__default["default"].createElement("div", {
    className: "mark"
  }));
};

var index = withCheckHOC["default"](Toggle);

exports["default"] = index;
