// @cred/neopop-web@1.0.0/lib/components/Helpers/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../_virtual/_rollupPluginBabelHelpers.js');
var styled = require('styled-components');
require('../../primitives/index.js');
var React = require('react');
var types = require('./types.js');
var index = require('../../utils/index.js');
var colors = require('../../primitives/colors.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

function _interopNamespace(e) {
    if (e && e.__esModule) return e;
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n["default"] = e;
    return Object.freeze(n);
}

var styled__default = /*#__PURE__*/_interopDefaultLegacy(styled);
var React__namespace = /*#__PURE__*/_interopNamespace(React);

var _RotateDegrees;
var Row = styled__default["default"].div.withConfig({
  displayName: "Helpers__Row",
  componentId: "sc-l5dlkf-0"
})(["display:flex;flex-direction:row;&.h-center{align-items:center;}&.v-center{justify-content:center;}&.v-justify{justify-content:space-between;}"]);
var Column = styled__default["default"].div.withConfig({
  displayName: "Helpers__Column",
  componentId: "sc-l5dlkf-1"
})(["display:flex;flex-direction:column;&.v-center{align-items:center;}&.h-center{justify-content:center;}"]);
var PageContainer = styled__default["default"].div.withConfig({
  displayName: "Helpers__PageContainer",
  componentId: "sc-l5dlkf-2"
})(["background-color:", ";min-height:calc(100vh - constant(safe-area-inset-top));min-height:calc(100vh - env(safe-area-inset-top));width:100vw;padding:0px 15px 0px 30px;padding-bottom:constant(safe-area-inset-bottom);padding-bottom:env(safe-area-inset-bottom);display:flex;flex-direction:column;"], colors.mainColors.black);
var HorizontalDivider = styled__default["default"].div.withConfig({
  displayName: "Helpers__HorizontalDivider",
  componentId: "sc-l5dlkf-3"
})(["background-color:", ";margin:0px;width:100%;height:1px;position:relative;"], function (_ref) {
  var color = _ref.color;
  return color !== null && color !== void 0 ? color : index.hexToRGBA(colors.mainColors.white, 0.1);
});
var HorizontalSpacer = styled__default["default"].div.withConfig({
  displayName: "Helpers__HorizontalSpacer",
  componentId: "sc-l5dlkf-4"
})(["margin-bottom:", ";"], function (_ref2) {
  var n = _ref2.n;
  return "calc(".concat(n, " * 5px)");
});
var VerticalSpacer = styled__default["default"].div.withConfig({
  displayName: "Helpers__VerticalSpacer",
  componentId: "sc-l5dlkf-5"
})(["margin-right:", ";"], function (_ref3) {
  var n = _ref3.n;
  return "calc(".concat(n, " * 5px)");
});
var RotateDegrees = (_RotateDegrees = {}, _rollupPluginBabelHelpers.defineProperty(_RotateDegrees, types.Directions.NORTH, '-90deg'), _rollupPluginBabelHelpers.defineProperty(_RotateDegrees, types.Directions.SOUTH, '90deg'), _rollupPluginBabelHelpers.defineProperty(_RotateDegrees, types.Directions.EAST, '0deg'), _rollupPluginBabelHelpers.defineProperty(_RotateDegrees, types.Directions.WEST, '180deg'), _RotateDegrees);
var Chevron = function Chevron(_ref4) {
  var _ref4$color = _ref4.color,
      color = _ref4$color === void 0 ? colors.mainColors.white : _ref4$color,
      _ref4$direction = _ref4.direction,
      direction = _ref4$direction === void 0 ? types.Directions.EAST : _ref4$direction;
  return /*#__PURE__*/React__namespace.createElement("svg", {
    style: {
      transform: "rotate(".concat(RotateDegrees[direction !== null && direction !== void 0 ? direction : types.Directions.EAST], ")")
    },
    width: "8",
    height: "11",
    viewBox: "0 0 8 11",
    fill: "none"
  }, /*#__PURE__*/React__namespace.createElement("path", {
    d: "M2 9.5L6 5.5L2 1.5",
    stroke: color,
    strokeWidth: "1.5",
    strokeLinecap: "square"
  }));
};
var Pointer = function Pointer(_ref5) {
  var _ref5$color = _ref5.color,
      color = _ref5$color === void 0 ? colors.mainColors.white : _ref5$color,
      _ref5$direction = _ref5.direction,
      direction = _ref5$direction === void 0 ? types.Directions.EAST : _ref5$direction,
      _ref5$size = _ref5.size,
      size = _ref5$size === void 0 ? 'medium' : _ref5$size,
      style = _ref5.style;
  return size === 'small' ? /*#__PURE__*/React__namespace.createElement("svg", {
    style: _rollupPluginBabelHelpers.objectSpread2({
      transform: "rotate(".concat(RotateDegrees[direction !== null && direction !== void 0 ? direction : types.Directions.EAST], ")")
    }, style),
    width: "13",
    height: "9",
    viewBox: "0 0 13 9",
    fill: "none"
  }, /*#__PURE__*/React__namespace.createElement("path", {
    d: "M1 3.74996H0.25V5.24996H1V3.74996ZM1 5.24996H11.9997V3.74996H1V5.24996ZM8.37874 0.5C8.37874 3.09998 10.4246 5.25 12.9997 5.25V3.75C11.299 3.75 9.87874 2.31829 9.87874 0.5L8.37874 0.5ZM12.9997 3.75C10.4246 3.75 8.37865 5.89999 8.37865 8.5H9.87865C9.87865 6.68173 11.299 5.25 12.9997 5.25V3.75Z",
    fill: color
  })) : /*#__PURE__*/React__namespace.createElement("svg", {
    style: _rollupPluginBabelHelpers.objectSpread2({
      transform: "rotate(".concat(RotateDegrees[direction !== null && direction !== void 0 ? direction : types.Directions.EAST], ")")
    }, style),
    width: "21",
    height: "9",
    viewBox: "0 0 21 9",
    fill: "none"
  }, /*#__PURE__*/React__namespace.createElement("path", {
    d: "M1 3.74996H0.25L0.25 5.24996L1 5.24996L1 3.74996ZM1 5.24996L20 5.24996V3.74996L1 3.74996L1 5.24996ZM16.379 0.5C16.379 3.09998 18.4249 5.25 21 5.25V3.75C19.2993 3.75 17.879 2.3183 17.879 0.5L16.379 0.5ZM21 3.75C18.4249 3.75 16.3789 5.89999 16.3789 8.5H17.8789C17.8789 6.68172 19.2993 5.25 21 5.25V3.75Z",
    fill: color
  }));
};
var Cross = function Cross(_ref6) {
  var _ref6$color = _ref6.color,
      color = _ref6$color === void 0 ? colors.mainColors.white : _ref6$color,
      onClick = _ref6.onClick;
  return /*#__PURE__*/React__namespace.createElement("svg", {
    width: "12",
    height: "12",
    viewBox: "0 0 12 12",
    fill: "none",
    onClick: onClick
  }, /*#__PURE__*/React__namespace.createElement("path", {
    d: "M2 10L6 6L2 2",
    stroke: color,
    strokeWidth: "1.5",
    strokeLinecap: "square"
  }), /*#__PURE__*/React__namespace.createElement("path", {
    d: "M10 10L6 6L10 2",
    stroke: color,
    strokeWidth: "1.5",
    strokeLinecap: "square"
  }));
};

exports.Chevron = Chevron;
exports.Column = Column;
exports.Cross = Cross;
exports.HorizontalDivider = HorizontalDivider;
exports.HorizontalSpacer = HorizontalSpacer;
exports.PageContainer = PageContainer;
exports.Pointer = Pointer;
exports.Row = Row;
exports.VerticalSpacer = VerticalSpacer;
