// @cred/neopop-web@1.0.0/lib-esm/components/Header/index.js
import { extends as _extends } from '../../_virtual/_rollupPluginBabelHelpers.js';
import React__default from 'react';
import '../../primitives/index.js';
import Back from '../Back/index.js';
import { Column } from '../Helpers/index.js';
import Typography from '../Typography/index.js';
import { typographyGuide } from '../../primitives/typography.js';
import { mainColors } from '../../primitives/colors.js';
import { FontOpacity } from '../../primitives/opacity.js';

var Header = function Header(props) {
  var header = typographyGuide.header;
  var heading = props.heading,
      _props$textStyle = props.textStyle,
      textStyle = _props$textStyle === void 0 ? {
    heading: header.heading,
    description: header.description
  } : _props$textStyle,
      color = props.color,
      description = props.description,
      onBackClick = props.onBackClick;
  var fallbackColor = color !== null && color !== void 0 ? color : mainColors.white;
  return /*#__PURE__*/React__default.createElement(Column, null, /*#__PURE__*/React__default.createElement(Back, {
    onClick: onBackClick,
    color: fallbackColor
  }), heading ? /*#__PURE__*/React__default.createElement("div", {
    style: {
      paddingTop: '10px'
    }
  }, /*#__PURE__*/React__default.createElement(Typography, _extends({}, textStyle.heading, {
    style: {
      maxWidth: '75%',
      opacity: FontOpacity.HEADING
    },
    color: fallbackColor
  }), heading)) : '', description ? /*#__PURE__*/React__default.createElement("div", {
    style: {
      paddingTop: '5px'
    }
  }, /*#__PURE__*/React__default.createElement(Typography, _extends({}, textStyle.description, {
    style: {
      maxWidth: '75%',
      opacity: FontOpacity.BODY_TEXT
    },
    color: fallbackColor
  }), description)) : '');
};

export { Header as default };
