// @cred/neopop-web@1.0.0/lib/components/Toast/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var _rollupPluginBabelHelpers = require('../../_virtual/_rollupPluginBabelHelpers.js');
var React = require('react');
var ToastPortal = require('./components/ToastPortal.js');

function _interopDefaultLegacy (e) { return e && typeof e === 'object' && 'default' in e ? e : { 'default': e }; }

var React__default = /*#__PURE__*/_interopDefaultLegacy(React);

var GlobalStore = /*#__PURE__*/function () {
  function GlobalStore() {
    _rollupPluginBabelHelpers.classCallCheck(this, GlobalStore);

    _rollupPluginBabelHelpers.defineProperty(this, "func", void 0);

    this.func = [];
  }

  _rollupPluginBabelHelpers.createClass(GlobalStore, [{
    key: "on",
    value: function on(listener) {
      this.func.push(listener);
    }
  }, {
    key: "emit",
    value: function emit(options) {
      this.func.forEach(function (l) {
        return l(options);
      });
    }
  }]);

  return GlobalStore;
}();

var globalStore = new GlobalStore();
var showToast = function showToast(content, options) {
  globalStore.emit(_rollupPluginBabelHelpers.objectSpread2(_rollupPluginBabelHelpers.objectSpread2({}, options), {}, {
    content: content
  }));
};
var ToastContainer = function ToastContainer() {
  var toastRef = React.useRef();
  React.useEffect(function () {
    var listener = function listener(options) {
      var _toastRef$current;

      (_toastRef$current = toastRef.current) === null || _toastRef$current === void 0 ? void 0 : _toastRef$current.addToast(options);
    };

    globalStore.on(listener);
  }, []);
  return /*#__PURE__*/React__default["default"].createElement(ToastPortal.ToastPortal, {
    ref: toastRef
  });
};

exports.ToastContainer = ToastContainer;
exports.showToast = showToast;
