// @cred/neopop-web@1.0.0/lib-esm/components/InputField/index.js
import { objectWithoutProperties as _objectWithoutProperties, extends as _extends } from '../../_virtual/_rollupPluginBabelHelpers.js';
import React__default from 'react';
import Typography from '../Typography/index.js';
import { Container, Label, Input } from './styles.js';
import useScrollIntoView from '../../hooks/useScrollIntoView.js';
import useAutoFocus from '../../hooks/useAutoFocus.js';
import '../../primitives/index.js';
import { isEmpty } from '../../utils/index.js';
import { typographyGuide } from '../../primitives/typography.js';
import { colorGuide } from '../../primitives/colors.js';

var _excluded = ["id", "label", "colorMode", "isDisabled", "inputRef", "scrollIntoView", "autoFocus", "value", "colorConfig", "textStyle", "hasError", "errorMessage"];

var InputField = function InputField(props) {
  var _textStyle$label, _textStyle$input, _textStyle$input2;

  var id = props.id,
      label = props.label,
      colorMode = props.colorMode,
      isDisabled = props.isDisabled,
      inputRef = props.inputRef,
      scrollIntoView = props.scrollIntoView,
      autoFocus = props.autoFocus,
      value = props.value,
      colorConfig = props.colorConfig,
      textStyle = props.textStyle,
      hasError = props.hasError,
      _props$errorMessage = props.errorMessage,
      errorMessage = _props$errorMessage === void 0 ? 'error message' : _props$errorMessage,
      propsToFwd = _objectWithoutProperties(props, _excluded);

  var customRef = React__default.useRef(null);
  var inputFields = typographyGuide.inputFields;
  useScrollIntoView({
    scrollIntoView: scrollIntoView,
    ref: customRef
  });
  useAutoFocus({
    autoFocus: autoFocus,
    ref: customRef,
    currentValue: value
  });

  var getLabelColor = function getLabelColor() {
    if (hasError) {
      return colorConfig === null || colorConfig === void 0 ? void 0 : colorConfig.errorColor;
    }

    return colorConfig === null || colorConfig === void 0 ? void 0 : colorConfig.labelColor;
  };

  var labelText = function labelText() {
    if (hasError) {
      return errorMessage;
    }

    return label;
  };

  var defaultColorConfig = colorGuide[colorMode === 'light' ? 'lightComponents' : 'darkComponents'].inputFields;
  var inputFieldColorConfig = isEmpty(colorConfig) ? defaultColorConfig : colorConfig !== null && colorConfig !== void 0 ? colorConfig : defaultColorConfig;
  return /*#__PURE__*/React__default.createElement(Container, null, label ? /*#__PURE__*/React__default.createElement(Label, {
    htmlFor: id
  }, /*#__PURE__*/React__default.createElement(Typography, _extends({}, (_textStyle$label = textStyle === null || textStyle === void 0 ? void 0 : textStyle.label) !== null && _textStyle$label !== void 0 ? _textStyle$label : inputFields.label, {
    color: getLabelColor()
  }), labelText())) : '', /*#__PURE__*/React__default.createElement(Typography, (_textStyle$input = textStyle === null || textStyle === void 0 ? void 0 : textStyle.input) !== null && _textStyle$input !== void 0 ? _textStyle$input : inputFields.text, /*#__PURE__*/React__default.createElement(Input, _extends({
    ref: function ref(_ref) {
      customRef.current = inputRef ? inputRef.current = _ref : _ref;
    },
    value: value,
    id: id,
    disabled: isDisabled,
    textColor: inputFieldColorConfig === null || inputFieldColorConfig === void 0 ? void 0 : inputFieldColorConfig.textColor,
    caretColor: inputFieldColorConfig === null || inputFieldColorConfig === void 0 ? void 0 : inputFieldColorConfig.caretColor,
    placeholderColor: inputFieldColorConfig === null || inputFieldColorConfig === void 0 ? void 0 : inputFieldColorConfig.placeholderColor,
    text: (_textStyle$input2 = textStyle === null || textStyle === void 0 ? void 0 : textStyle.input) !== null && _textStyle$input2 !== void 0 ? _textStyle$input2 : inputFields.text
  }, propsToFwd))));
};

export { InputField as default };
