// @cred/neopop-web@1.0.0/lib-esm/components/Helpers/index.js
import { defineProperty as _defineProperty, objectSpread2 as _objectSpread2 } from '../../_virtual/_rollupPluginBabelHelpers.js';
import styled from 'styled-components';
import '../../primitives/index.js';
import * as React from 'react';
import { Directions } from './types.js';
import { hexToRGBA } from '../../utils/index.js';
import { mainColors } from '../../primitives/colors.js';

var _RotateDegrees;
var Row = styled.div.withConfig({
  displayName: "Helpers__Row",
  componentId: "sc-l5dlkf-0"
})(["display:flex;flex-direction:row;&.h-center{align-items:center;}&.v-center{justify-content:center;}&.v-justify{justify-content:space-between;}"]);
var Column = styled.div.withConfig({
  displayName: "Helpers__Column",
  componentId: "sc-l5dlkf-1"
})(["display:flex;flex-direction:column;&.v-center{align-items:center;}&.h-center{justify-content:center;}"]);
var PageContainer = styled.div.withConfig({
  displayName: "Helpers__PageContainer",
  componentId: "sc-l5dlkf-2"
})(["background-color:", ";min-height:calc(100vh - constant(safe-area-inset-top));min-height:calc(100vh - env(safe-area-inset-top));width:100vw;padding:0px 15px 0px 30px;padding-bottom:constant(safe-area-inset-bottom);padding-bottom:env(safe-area-inset-bottom);display:flex;flex-direction:column;"], mainColors.black);
var HorizontalDivider = styled.div.withConfig({
  displayName: "Helpers__HorizontalDivider",
  componentId: "sc-l5dlkf-3"
})(["background-color:", ";margin:0px;width:100%;height:1px;position:relative;"], function (_ref) {
  var color = _ref.color;
  return color !== null && color !== void 0 ? color : hexToRGBA(mainColors.white, 0.1);
});
var HorizontalSpacer = styled.div.withConfig({
  displayName: "Helpers__HorizontalSpacer",
  componentId: "sc-l5dlkf-4"
})(["margin-bottom:", ";"], function (_ref2) {
  var n = _ref2.n;
  return "calc(".concat(n, " * 5px)");
});
var VerticalSpacer = styled.div.withConfig({
  displayName: "Helpers__VerticalSpacer",
  componentId: "sc-l5dlkf-5"
})(["margin-right:", ";"], function (_ref3) {
  var n = _ref3.n;
  return "calc(".concat(n, " * 5px)");
});
var RotateDegrees = (_RotateDegrees = {}, _defineProperty(_RotateDegrees, Directions.NORTH, '-90deg'), _defineProperty(_RotateDegrees, Directions.SOUTH, '90deg'), _defineProperty(_RotateDegrees, Directions.EAST, '0deg'), _defineProperty(_RotateDegrees, Directions.WEST, '180deg'), _RotateDegrees);
var Chevron = function Chevron(_ref4) {
  var _ref4$color = _ref4.color,
      color = _ref4$color === void 0 ? mainColors.white : _ref4$color,
      _ref4$direction = _ref4.direction,
      direction = _ref4$direction === void 0 ? Directions.EAST : _ref4$direction;
  return /*#__PURE__*/React.createElement("svg", {
    style: {
      transform: "rotate(".concat(RotateDegrees[direction !== null && direction !== void 0 ? direction : Directions.EAST], ")")
    },
    width: "8",
    height: "11",
    viewBox: "0 0 8 11",
    fill: "none"
  }, /*#__PURE__*/React.createElement("path", {
    d: "M2 9.5L6 5.5L2 1.5",
    stroke: color,
    strokeWidth: "1.5",
    strokeLinecap: "square"
  }));
};
var Pointer = function Pointer(_ref5) {
  var _ref5$color = _ref5.color,
      color = _ref5$color === void 0 ? mainColors.white : _ref5$color,
      _ref5$direction = _ref5.direction,
      direction = _ref5$direction === void 0 ? Directions.EAST : _ref5$direction,
      _ref5$size = _ref5.size,
      size = _ref5$size === void 0 ? 'medium' : _ref5$size,
      style = _ref5.style;
  return size === 'small' ? /*#__PURE__*/React.createElement("svg", {
    style: _objectSpread2({
      transform: "rotate(".concat(RotateDegrees[direction !== null && direction !== void 0 ? direction : Directions.EAST], ")")
    }, style),
    width: "13",
    height: "9",
    viewBox: "0 0 13 9",
    fill: "none"
  }, /*#__PURE__*/React.createElement("path", {
    d: "M1 3.74996H0.25V5.24996H1V3.74996ZM1 5.24996H11.9997V3.74996H1V5.24996ZM8.37874 0.5C8.37874 3.09998 10.4246 5.25 12.9997 5.25V3.75C11.299 3.75 9.87874 2.31829 9.87874 0.5L8.37874 0.5ZM12.9997 3.75C10.4246 3.75 8.37865 5.89999 8.37865 8.5H9.87865C9.87865 6.68173 11.299 5.25 12.9997 5.25V3.75Z",
    fill: color
  })) : /*#__PURE__*/React.createElement("svg", {
    style: _objectSpread2({
      transform: "rotate(".concat(RotateDegrees[direction !== null && direction !== void 0 ? direction : Directions.EAST], ")")
    }, style),
    width: "21",
    height: "9",
    viewBox: "0 0 21 9",
    fill: "none"
  }, /*#__PURE__*/React.createElement("path", {
    d: "M1 3.74996H0.25L0.25 5.24996L1 5.24996L1 3.74996ZM1 5.24996L20 5.24996V3.74996L1 3.74996L1 5.24996ZM16.379 0.5C16.379 3.09998 18.4249 5.25 21 5.25V3.75C19.2993 3.75 17.879 2.3183 17.879 0.5L16.379 0.5ZM21 3.75C18.4249 3.75 16.3789 5.89999 16.3789 8.5H17.8789C17.8789 6.68172 19.2993 5.25 21 5.25V3.75Z",
    fill: color
  }));
};
var Cross = function Cross(_ref6) {
  var _ref6$color = _ref6.color,
      color = _ref6$color === void 0 ? mainColors.white : _ref6$color,
      onClick = _ref6.onClick;
  return /*#__PURE__*/React.createElement("svg", {
    width: "12",
    height: "12",
    viewBox: "0 0 12 12",
    fill: "none",
    onClick: onClick
  }, /*#__PURE__*/React.createElement("path", {
    d: "M2 10L6 6L2 2",
    stroke: color,
    strokeWidth: "1.5",
    strokeLinecap: "square"
  }), /*#__PURE__*/React.createElement("path", {
    d: "M10 10L6 6L10 2",
    stroke: color,
    strokeWidth: "1.5",
    strokeLinecap: "square"
  }));
};

export { Chevron, Column, Cross, HorizontalDivider, HorizontalSpacer, PageContainer, Pointer, Row, VerticalSpacer };
